### PR TITLE
Pretrain gans

### PR DESCRIPTION
**Pull request by starcksophie**
Added:
• pre-training for adversarial training
• a benchmark for the generator
• a comparison between the sole generator, a gan setup, a gan setup + pre-training
•  safe multithreading for data generators.

### CODE REVIEW

File: fastmri_recon/helpers/adversarial_training.py [hunk 1]
```diff
@@ -8,7 +9,6 @@
 from .keras_utils import wasserstein_loss, mean_output, discriminator_accuracy
 from .utils import keras_ssim, keras_psnr
 
-
```
Comment (zaccharieramzi) [inline fastmri_recon/helpers/adversarial_training.py]: try to avoid whiteline changes (here and elsewhere)

File: fastmri_recon/helpers/adversarial_training.py [hunk 2]
```diff
@@ -1,3 +1,4 @@
+import tensorflow as tf
```
Comment (zaccharieramzi) [inline fastmri_recon/helpers/adversarial_training.py]: for imports try to respect the alphabetical order (same for elsewhere)

File: fastmri_recon/helpers/adversarial_training.py [hunk 3]
```diff
@@ -117,7 +118,11 @@ def adversarial_training_loop(
         use_multiprocessing=False,
         max_queue_size=10,
         include_d_metrics=False,
+        gen_pre_training_steps=0,
+        pre_training_callbacks=None,
+        # inner_training_gen=1
```
Comment (zaccharieramzi) [inline fastmri_recon/helpers/adversarial_training.py]: I think this is legacy

File: fastmri_recon/helpers/adversarial_training.py [hunk 4]
```diff
@@ -126,7 +131,7 @@ def adversarial_training_loop(
         callbacks,
         n_epochs=n_epochs,
         n_batches=n_batches,
-        include_d_metrics=include_d_metrics,
+        include_d_metrics=include_d_metrics
```
Comment (zaccharieramzi) [inline fastmri_recon/helpers/adversarial_training.py]: try to avoid these unnecessary changes

File: fastmri_recon/helpers/adversarial_training.py [hunk 5]
```diff
@@ -138,6 +143,14 @@ def adversarial_training_loop(
         use_sequence_api=use_sequence_api,
     )
 
+    # generator pre-training
+    pretraining_history = g.fit_generator(
+                                train_gen,
```
Comment (zaccharieramzi) [inline fastmri_recon/helpers/adversarial_training.py]: here you don't need to indent so much

File: fastmri_recon/helpers/adversarial_training.py [hunk 6]
```diff
@@ -194,3 +207,4 @@ def adversarial_training_loop(
                 epoch_logs[l] = o
         callbacks.on_epoch_end(epoch, epoch_logs)
     callbacks._call_end_hook('train')
+    return pretraining_history
```
Comment (zaccharieramzi) [inline fastmri_recon/helpers/adversarial_training.py]: don't forget the newline at the end of the file (same for elsewhere).
Maybe your editor isn't configured to add it at save, if not try to make it so.

File: fastmri_recon/data/test_generators.py [hunk 7]
```diff
@@ -38,8 +41,53 @@ def flow_random_shapes(self,):
             mask_batch = mask_batch[..., 0]
             yield (kspaces, mask_batch), images
 
+    @threadsafe_generator
     def flow_z_filled_random_shapes(self,):
         random_shapes_gen = self.flow_random_shapes()
-        for (kspaces, _), images in random_shapes_gen:
+        for (kspaces, mask_batch), images in random_shapes_gen:
+            z_filled = zero_filled_recon(kspaces[..., 0], crop=False)[..., None]
+            yield z_filled, images, mask_batch
```
Comment (zaccharieramzi) [inline fastmri_recon/data/test_generators.py]: why return the mask too now?

File: fastmri_recon/data/test_generators.py [hunk 8]
```diff
@@ -38,8 +41,53 @@ def flow_random_shapes(self,):
...
+
+class CifarGenerator:
+    def __init__(self, af, size, data, max_size, batch_size=1):
+        self.af = af
+        self.data = data
+        self.size = size
+        self.batch_size = batch_size
+        self.im_shape = (batch_size, size, size, 1)
+        self.index = 0
+        self.max_size = max_size
```
Comment (zaccharieramzi) [inline fastmri_recon/data/test_generators.py]: isn't `max_size` a function of `data` ? same for `size`

File: fastmri_recon/data/test_generators.py [hunk 9]
```diff
@@ -38,8 +41,53 @@ def flow_random_shapes(self,):
...
+            fourier_mask = np.repeat(mask.astype(np.float), self.size, axis=0)
+            mask_batch = np.repeat(fourier_mask[None, ...], len(kspaces), axis=0)[..., None]
+            kspaces *= mask_batch
+            mask_batch = mask_batch[..., 0]
+            yield (kspaces, mask_batch), images
+
+    @threadsafe_generator
+    def flow_z_filled_images(self,):
+        random_shapes_gen = self.flow_images()
+        for (kspaces, mask_batch), images in random_shapes_gen:
```
Comment (zaccharieramzi) [inline fastmri_recon/data/test_generators.py]: here you don't need `mask_batch` right?

File: fastmri_recon/data/test_generators.py [hunk 10]
```diff
@@ -38,8 +41,53 @@ def flow_random_shapes(self,):
...
+            images = np.empty(self.im_shape)
+            kspaces = np.empty(self.im_shape)
+            if self.index == self.max_size:
+                self.reset()
+            for i in range(self.batch_size):
+                image = self.data[self.index]
+                images[i, ..., 0] = image
+                
+                image = image.astype('float32')
+                image /= 255
```
Comment (zaccharieramzi) [inline fastmri_recon/data/test_generators.py]: I just realized that the images aren't centered around 0!
Can you confirm and if so add : `image -= 0.5`?

File: fastmri_recon/helpers/adversarial_training.py (hunk 3 above)
Comment (starcksophie) [reply fastmri_recon/helpers/adversarial_training.py]: These first two parameters are actually necessary for the pretraining, but I will remove the commented line.

File: fastmri_recon/data/test_generators.py (hunk 7 above)
Comment (starcksophie) [reply fastmri_recon/data/test_generators.py]: It was for testing purposes, my bad for leaving it there.

File: fastmri_recon/data/test_generators.py (hunk 8 above)
Comment (starcksophie) [reply fastmri_recon/data/test_generators.py]: Indeed

**Comment by zaccharieramzi** [comment]
Regarding the notebooks,
What I would like is to have:
- a notebook which pre-trains the generator but using the `adversarial_training_loop` so basically updating the `gan_unet_approach_random` notebook (this is the equivalent of your `gan_benchmark_generator_training` but within the adversarial framework),
- a notebook for validation where you have the numbers for different models trained adversarially (for now you just have one which is just pre-trained), for example `adversarial_training_validation` (this is the equivalent of your `gan_benchmark_generator`, but that can be used for any adversarially trained model and without the figures),
- if you want a notebook with a qualitative validation, i.e. including the figures.

For now we don't need to have the notebook for the experimentation between different strategies of pre-training for the GANs.

File: fastmri_recon/data/test_generators.py [hunk 11]
```diff
@@ -4,6 +4,7 @@
 from ..helpers.fourier import fft
 from ..helpers.reconstruction import zero_filled_recon
 from ..helpers.utils import gen_mask
+from ..helpers.threadsafe_gen import threadsafe_generator
```
Comment (zaccharieramzi) [inline fastmri_recon/data/test_generators.py]: Watch out for the order of the imports

File: fastmri_recon/data/test_generators.py [hunk 12]
```diff
@@ -14,6 +15,8 @@ def __init__(self, af, n_shapes, size, batch_size=1):
         self.batch_size = batch_size
         self.im_shape = (batch_size, size, size, 1)
 
+
```
Comment (zaccharieramzi) [inline fastmri_recon/data/test_generators.py]: watch out for this extra whiteline

File: fastmri_recon/helpers/adversarial_training.py [hunk 13]
```diff
@@ -44,6 +45,7 @@ def prepare_callbacks(g, d, d_on_g, callbacks, n_epochs=1, n_batches=1, include_
     # we only want to validate on the output of g
     val_out_labels = ['val_' + n for n in out_labels if g.name in n]
     callback_metrics = out_labels + val_out_labels
+
```
Comment (zaccharieramzi) [inline fastmri_recon/helpers/adversarial_training.py]: watch out this whiteline

File: fastmri_recon/helpers/adversarial_training.py [hunk 14]
```diff
@@ -168,7 +183,7 @@ def adversarial_training_loop(
                     d_outs_fake,
                     d_outs_real,
                     d_metrics_fake,
-                    d_metrics_real,
+                    d_metrics_real
```
Comment (zaccharieramzi) [inline fastmri_recon/helpers/adversarial_training.py]: watch out for this removed comma

File: fastmri_recon/data/test_generators.py [hunk 15]
```diff
@@ -38,8 +41,53 @@ def flow_random_shapes(self,):
...
         random_shapes_gen = self.flow_random_shapes()
         for (kspaces, _), images in random_shapes_gen:
             z_filled = zero_filled_recon(kspaces[..., 0], crop=False)[..., None]
             yield z_filled, images
+
+class DataGenerator:
+    def __init__(self, af, data, batch_size=1):
+        self.af = af
+        self.data = data
+        self.size = data.shape[1]
```
Comment (zaccharieramzi) [inline fastmri_recon/data/test_generators.py]: Here size is not very explicit since you only consider the first dimension of the image. We might have rectangular images soon. Also you probably don't need this since you have `im_shape` just under.

File: fastmri_recon/data/test_generators.py [hunk 16]
```diff
@@ -38,8 +41,53 @@ def flow_random_shapes(self,):
...
+        self.batch_size = batch_size
+        self.im_shape = (batch_size, data.shape[1], data.shape[1], 1)
```
Comment (zaccharieramzi) [inline fastmri_recon/data/test_generators.py]: `self.im_shape = (batch_size, data.shape[1], data.shape[2], 1)`

**Comment by zaccharieramzi** [comment]
Regarding the notebooks:
- in the `gan_unet_approach_random`, remove all the cells after the one printing the run id (they belong to a validation notebook). Also remove everything related to plotting. The idea is to make the notebook as slim as possible, an if someone wants to look a bit more into it they can always plot themselves.
- Regarding the `adversarial_training_validation`, take the other validation notebooks for example in order to have a similar notebook but for adversarially trained networks. In particular, one should be able to provide a list of parameters and have the corresponding models' performance metrics all listed in a table.

File: fastmri_recon/helpers/adversarial_training.py [hunk 17]
```diff
@@ -117,7 +118,10 @@ def adversarial_training_loop(
         use_multiprocessing=False,
         max_queue_size=10,
         include_d_metrics=False,
+        gen_pre_training_steps=0,
+        pre_training_callbacks=None,
     ):
+
```
Comment (zaccharieramzi) [inline fastmri_recon/helpers/adversarial_training.py]: watch out for this whiteline
